lxd/storage/drivers/generic.go

```diff
@@ -0,0 +1,288 @@
+package drivers
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/lxc/lxd/lxd/migration"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/rsync"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/ioprogress"
+	log "github.com/lxc/lxd/shared/log15"
+)
+
+// genericCopyVolume copies a volume and its snapshots using a non-optimized method.
+func genericCopyVolume(d Driver, applyQuota func(vol Volume) (func(), error), vol Volume, srcVol Volume, srcSnapshots []Volume, op *operations.Operation) error {
```

Comment:
@stgraber just a nit really, but the `applyQuota` function argument here could perhaps just be called `preCopyHook` because we as inside the function we don't actually know that the function passed will apply the quota. It could be used in the future for another sort of pre copy action and then the naming might not make sense.

Reply:
So `preCopyHook` I think would be somewhat confusing because you may expect it to also be run for snapshots whereas quotas really only apply to the main volume.

Reply:
@stgraber yes that makes sense, although having a mention of "quota" when this function doesn't perform any quota actions might confuse in the future. How about `initVolume` as there is only one volume being copied the "Volume" in `genericCopyVolume`.

Reply:
@stgraber although there are srcSnapshots passed as Volumes too, so maybe thats confusing too. I guess either way it needs a comment in the function description explaining that this function (be it `initVolume` or `applyQuota`) is only run on the primary volume and not the snapshots.

Reply:
Got a commit doing that in my local branch, will be part of the next batch of storage changes from me.